docs/architecture/adr-001-node-key-refactoring.md

```diff
@@ -0,0 +1,101 @@
+# ADR ADR-001: Node Key Refactoring
+
+## Changelog
+
+- 2022-10-31: First draft
+
+## Status
+
+Proposed
+
+## Context
+
+The original key format of IAVL nodes is a hash of the node. It does not take advantage of data locality on disk. Nodes are stored in random locations on disk due to the random hash value, so it needs to scan the disk to find the corresponding node which can be very inefficient.
+
+The `orphans` are used to manage node removal in the current design and allow deletion of removed nodes for the specific version from the disk through the `DeleteVersion` API. It needs to track every time when updating the tree and also requires extra storage to store `orphans`, but there are not many use cases of `DeleteVersion`. There are two use cases:
+
+1. Rollback of the tree to a previous version
+2. Remove unnecessary old nodes
+
+## Decision
+
+- Use the sequenced integer ID as a node key like `bigendian(nodeKey)` format.
+- Remove the `leftHash` and `rightHash` fields, and instead store `hash` field.
+- Remove the `orphans` from the tree.
+
+Theoretically, we can also remove the `version` field in the node structure but it leads to breaking the ics23 proof mechanism. We will revisit it later.
+
+New node structure
+
+```go
+type Node struct {
+	key           []byte
+	value         []byte
+	hash          []byte    // keep this field in the storage
+	nodeKey       int64     // new field, use as a node key
+	leftNodeKey   int64     // new field, need to store in the storage
```

Comment:
If we are using nonce as `nodeKey`, how are we getting data locality when traversing or writing the tree to disk?

I thought that the whole idea with the path encoded as a key was that writes would be sequential, contributing to lower frequency of compactions.

Reply:
Without pruning, the nodes are all immutable, the `nodeKey` is sequential, so all the new nodes are written sequentially.

Reply:
I know the tree path describes the data locality well. The problem is how to keep the path when rotating the tree.

Reply:
we need to split the `leftNodeKey` into two fields, `leftNodeVersion` and `leftNodeNonce`, right?

Reply:
Aren't we need to split the `leftNodeKey` into `leftNodeVersion` and `leftNodeNonce`, likewise for `rightNodeKey`.

Reply:
right, I updated the node struct